File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        '''
        The Subtree can give me its depth, so i dont need a helper
        Every child, children subtree can be bubbled up until we reach parent. So pre order Left -> right-> Node
        '''

        if not root:
            return 0
        
        left_depth = self.maxDepth(root.left)
        right_depth = self.maxDepth(root.right)

        return 1 + max(left_depth,right_depth)
        

        
        
        

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        
       return dfs(root,0);
    }

    public int dfs(TreeNode node, int cur_count){

        if(node == null){
            return cur_count;
        }

        cur_count++;

        int left = dfs(node.left,cur_count);
        int right = dfs(node.right,cur_count);

        return Math.max(left,right);

    }
}